compiler/grainformat/comment_utils.re

```diff
@@ -0,0 +1,564 @@
+};
+
+let get_comments_between_locs =
+    (~loc1: Location.t, ~loc2: Location.t, comments: list(Parsetree.comment)) => {
+  let (_, stmt_end_line, stmt_end_char, _) =
+    Locations.get_raw_pos_info(loc1.loc_end);
+  let (_, stmt_start_line, stm_start_char, _) =
+    Locations.get_raw_pos_info(loc2.loc_start);
+
+  // invert the request to look inside the location in the gap
+
+  let start_loc: Lexing.position = {
+    pos_fname: "",
+    pos_lnum: stmt_end_line,
+    pos_bol: 0,
+    pos_cnum: stmt_end_char,
+  };
+  let end_loc: Lexing.position = {
+    pos_fname: "",
+    pos_lnum: stmt_start_line,
+    pos_bol: 0,
+    pos_cnum: stm_start_char,
+  };
+
+  let location: Location.t = {
+    loc_start: start_loc,
+    loc_end: end_loc,
+    loc_ghost: false,
+  };
+  get_comments_inside_location(~location, comments);
+};
+
+let rec get_comments_on_line_end =
+        (line: int, char: int, comments: list(Parsetree.comment)) =>
+  switch (comments) {
+  | [] => []
+  | [c, ...remaining_comments] =>
+    let c_loc: Location.t = Locations.get_comment_loc(c);
+    let (_, cmtline, cmtchar, _) =
+      Locations.get_raw_pos_info(c_loc.loc_start);
+
+    if (cmtline > line) {
+      []; // can stop early as there will be no more
+    } else if (cmtline == line && cmtchar >= char) {
+      [c, ...get_comments_on_line_end(line, char, List.tl(comments))];
+    } else {
+      get_comments_on_line_end(line, char, List.tl(comments));
+    };
+  };
+
+let rec get_comments_on_line_start = (line: int, char: int, comments) =>
+  switch (comments) {
+  | [] => []
+  | [c, ...remaining_comments] =>
+    let c_loc: Location.t = Locations.get_comment_loc(c);
+    let (_, cmtline, cmtchar, _) =
+      Locations.get_raw_pos_info(c_loc.loc_start);
+
+    if (cmtline > line) {
+      []; // can stop early as there will be no more
+    } else if (cmtline == line && cmtchar <= char) {
+      [c, ...get_comments_on_line_start(line, char, List.tl(comments))];
+    } else {
+      get_comments_on_line_start(line, char, List.tl(comments));
+    };
+  };
+
+let get_comments_to_end_of_line =
+    (~location: Location.t, comments: list(Parsetree.comment)) => {
+  let (_, stmt_end_line, stmt_end_char, _) =
+    Locations.get_raw_pos_info(location.loc_end);
+
+  get_comments_on_line_end(stmt_end_line, stmt_end_char, comments);
+};
+
+let comment_to_doc = (comment: Parsetree.comment) => {
+  let comment_string = Comments.get_comment_source(comment);
+  let newline =
+    switch (comment) {
+    | Line(_)
+    | Shebang(_) => Doc.hardLine
+    | Doc(_) => Doc.hardLine
+    | _ => Doc.nil
+    };
+  Doc.concat([Doc.text(String.trim(comment_string)), newline]);
+};
+
+let nobreak_comment_to_doc = (comment: Parsetree.comment) => {
+  let comment_string = Comments.get_comment_source(comment);
+
+  Doc.text(String.trim(comment_string));
+};
+
+let get_after_brace_comments =
+    (
+      loc: Location.t,
+      comments: list(Parsetree.comment),
+      first: option(Location.t),
```

Comment:
Why isn't this an optional, named param?

Reply:
Nice, fixed.